sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -376,6 +386,23 @@ object ScalaReflection extends ScalaReflection {
           dataType = ObjectType(udt.getClass))
         Invoke(obj, "deserialize", ObjectType(udt.userClass), getPath :: Nil)
 
+      case t if isValueClass(t) =>
+        // nested value class is treated as its underlying type
+        // top level value class must be treated as a product
```

Comment:
This is not true, top level value class can be a single primitive type column

Reply:
Here what I mean by "treated as a product" is we must create a new instance for it in case we have `Dataset[Id]` for example (as in my [comment](https://github.com/apache/spark/pull/22309#discussion_r226142827) above). Seems like this is confusing? Should I reword it?